app/modules/tmr.c

```diff
@@ -65,21 +65,54 @@ typedef void (*alarm_timer_callback)(void *arg);
 static alarm_timer_callback alarm_timer_cb[NUM_TMR] = {alarm_timer_cb0,alarm_timer_cb1,alarm_timer_cb2,alarm_timer_cb3,alarm_timer_cb4,alarm_timer_cb5,alarm_timer_cb6};
 
 // Lua: delay( us )
-static int tmr_delay( lua_State* L )
+static int tmr_delay_us( lua_State* L )
 {
   s32 us;
   us = luaL_checkinteger( L, 1 );
   if ( us <= 0 )
     return luaL_error( L, "wrong arg range" );
-  unsigned sec = (unsigned)us / 1000000;
-  unsigned remain = (unsigned)us % 1000000;
+
+     os_delay_us( us );
+     WRITE_PERI_REG(0x60000914, 0x73); //Watchdog clear to aviod system reboot
+
+  return 0;  
+}
+
+// Lua: delay( ms )
+static int tmr_delay_ms( lua_State* L )
+{
+  s32 ms;
+  ms = luaL_checkinteger( L, 1 );
+  if ( ms <= 0 )
+    return luaL_error( L, "wrong arg range" );
+    
   int i = 0;
-  for(i=0;i<sec;i++){
-    os_delay_us( 1000000 );
-    WRITE_PERI_REG(0x60000914, 0x73);
-  }
-  if(remain>0)
-    os_delay_us( remain );
+  for(i=0;i<1000;i++) // 1ms == 1000us. Simple stupid and should work.
```

Comment:
If I ask it to sleep for 30ms, it will try to sleep for 30us, 1000 times, which is 30 ms. It would be better to do the opposite; sleep for 1000us 30 times, in order to skip a couple of evaluations of the for loop.

In fact, a better solution would be to calculate (with a simple division) exactly how many times you can sleep for a full second (which seems to be a sensible interval to reset the watchdog) and then sleep for the remainder. In the example i gave you, that would be 0 full seconds and then 30000us.

In the case of implementing the second solution, it would be exactly as implemented in the original `tmr_delay`.